File: Unsupervised-Learning/bike_shops_pca.R
library(tidyverse)
library(corrplot)
library(ggthemes)
library(gridExtra)

# Set working directory!

# Load data in memory
# -------------------

customers <- read.csv("customers_bikeshops.csv")
orders <- read.csv("orders.csv")
products <- read.csv("products_bikes.csv")

# Merge datasets
# --------------

orders_products <- inner_join(orders, products, by = c("product.id" = "bike.id"))
bike_data <- inner_join(customers, orders_products, by = c("bikeshop.id" = "customer.id"))

# Group and Reshape Data
# ----------------------

cluster_data <- group_by(bike_data, bikeshop.name, model) %>%
  summarize(quantity = sum(quantity), price = sum(price)) %>%
  mutate(sales = quantity * price) %>%
  select(bikeshop.name, model, sales) %>%
  dcast(... ~ model, value.var = "sales")

X <- select(cluster_data, -1)
y <- select(cluster_data, 1)

# Fill NAs with 0s (NAs imply zero sales)
X <- apply(X, 2, as.numeric)
X[is.na(X)] <- 0

# PCA
# ---

# Our goal with PCA here is to reduce the dimensionality
# of the data on the bike sales per bike shop.

# We set the scale parameter in the prcomp() function to true
# so that we don't have to do the transformations
# manually.

model <- prcomp(X, scale = TRUE)

# Visualize explained variance ratio and cumulative explained variance ratio

model.var <- model$sdev^2
explained_variance_ratio <- model.var / sum(model.var)

pca_df <- data.frame("n_components" = seq(ncol(model$x)), 
                     "explained_variance_ratio" = explained_variance_ratio,
                     "evr_cumulative" = cumsum(explained_variance_ratio))

explained_variance_p1 <- ggplot(pca_df, aes(x = n_components, y = explained_variance_ratio)) +
  geom_line() +
  scale_x_continuous(breaks = pca_df$n_components) +
  ggtitle("Proportions of Variance Explained by each Component") +
  xlab("Number of Components") +
  ylab('Explained Variance Ratio') +
  theme_tufte()

explained_variance_p2 <- ggplot(pca_df, aes(x = n_components, y = evr_cumulative)) +
  geom_line() +
  scale_x_continuous(breaks = pca_df$n_components) +
  ggtitle("Proportions of Variance Explained by each Component (Cumulative)") +
  xlab("Number of Components") +
  ylab('Cumulative Explained Variance Ratio') +
  theme_tufte()

grid.arrange(explained_variance_p1, explained_variance_p2)

# We now have 5 components which capture over 90% of the variance.

File: Unsupervised-Learning/bike_shops_pca.py
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import scale
from sklearn.decomposition import PCA


# Load data into memory
# ---------------------

customers = pd.read_csv("customers_bikeshops.csv")
orders = pd.read_csv("orders.csv", index_col=0)
products = pd.read_csv("products_bikes.csv")

# Merge datasets
# --------------

# product.id = bike.id
# bikeshop.id = customer.id

orders_products = pd.merge(orders, products, left_on="product.id", right_on="bike.id")
bike_data = pd.merge(customers, orders_products, left_on="bikeshop.id", right_on="customer.id")

# Make correlation plot
# ---------------------

corr = bike_data.corr()
sns.heatmap(corr, xticklabels=corr.columns, yticklabels=corr.columns)

# No strong correlations between variables that are not identical detected.

# Group data
# ----------

# Group by bikeshop and bike model
cluster_data = bike_data.groupby(["bikeshop.name", "model"]).\
    agg({"quantity": "sum", "price": "sum"}).reset_index()

# Compute sales
cluster_data["sales"] = cluster_data["quantity"] * cluster_data["price"]

# Select columns of interest
cluster_data = cluster_data[["bikeshop.name", "model", "sales"]]

# Pivot data
cluster_data = cluster_data.pivot(index="bikeshop.name", columns="model", values="sales").reset_index()

# Fill NaNs with 0s (NaN indicates no sales)
cluster_data = cluster_data.fillna(0)

# Normalize Scores
# ----------------

# Normalize data so that mean = 0 and standard deviation = 1

X = cluster_data.drop(["bikeshop.name"], axis=1)
y = cluster_data["bikeshop.name"]

normalized_data = scale(X)
sns.heatmap(normalized_data)

# PCA
# ---

# Our goal with PCA here is to reduce the dimensionality
# of the data on the bike sales per bike shop

model = PCA().fit(normalized_data)

# Visualize explained variance ratio and cumulative explained variance ratio

plt.subplot(1, 2, 1)
plt.plot(range(len(model.explained_variance_ratio_)), model.explained_variance_ratio_)
plt.xlabel('Number of Components')
plt.ylabel('Explained Variance Ratio')
plt.title('Proportions of Variance Explained by each Component')


plt.subplot(1, 2, 2)
plt.plot(range(len(model.explained_variance_ratio_)), np.cumsum(model.explained_variance_ratio_))
plt.xlabel('Number of Components')
plt.ylabel('Cumulative Explained Variance Ratio')
plt.show()

# We now have 5 components which capture over 90% of the variance.
